### PR TITLE
Revise Processing Model to separate initialization phase

### LINKED ISSUE #1597: More questions on rendering a graph
Look at the [algorithm for rendering a graph](https://webaudio.github.io/web-audio-api/#rendering-loop).

We have several steps:

1. Prepare the rendering. (We set frame to 0)
1. Process the control message queue.
1. Process a render quantum. (Substep 6 increments frame).
1. Perform a microtask checkpoint.
1. Return RenderResult.

Step 1 is a bit of problem if this algorithm is invoked for every render.  I think we're missing some kind of loop.  Perhaps step 5 should say we return RenderResult and go back to step 2.  I can't see that we ever want to run step 1 ever again.

### PR DESCRIPTION
**Pull request by hoch**
PTAL - fixes #1597.

***
<a href="https://pr-preview.s3.amazonaws.com/hoch/web-audio-api/pull/1629.html" title="Last updated on May 24, 2018, 7:39 PM GMT (4a62968)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WebAudio/web-audio-api/1629/1cf04b6...hoch:4a62968.html" title="Last updated on May 24, 2018, 7:39 PM GMT (4a62968)">Diff</a>

### CODE REVIEW

**Comment by hoch** [comment]
Not a problem. Once #1630 is merged, I'll rebase upon it.

File: index.bs [hunk 1]
```diff
@@ -10552,29 +10552,32 @@ of <a href="#rendering-a-graph">rendering a graph</a>.
...
 <a>rendering thread</a> has another task queue for microtasks for any microtask
 operation such as resolution of {{Promise}}s in the {{AudioWorkletGlobalScope}}.
 
 <div id="rendering-a-graph" algorithm="rendering a graph">
-	1. Prepare the rendering.
 
-		1. Let <var>Q<sub>rendering</sub></var> be an empty [=control message
-			queue=]. [=Atomically=] [=swap=]
-			<var>Q<sub>rendering</sub></var> and <var>Q</var>.
+	<div id="rendering-initialization" algorithm="initialize rendering loop">
```
Comment (rtoy) [inline index.bs]: I think this should be outside the "rendering a graph" algorithm.  At least the preview shows funny nested boxes which somehow doesn't look right.

File: index.bs [hunk 2]
```diff
@@ -10552,29 +10552,32 @@ of <a href="#rendering-a-graph">rendering a graph</a>.
...
+		Before the rendering loop starts, UA MUST perform the following steps once
+		for the initialization:
+
+		1. Let <var>Q<sub>rendering</sub></var> and <var>Q<sub>deferred</sub></var>
```
Comment (rtoy) [inline index.bs]: Qrendering gets initialized to empty in the rendering loop so there's no need to say it here.

File: index.bs [hunk 3]
```diff
@@ -10552,29 +10552,32 @@ of <a href="#rendering-a-graph">rendering a graph</a>.
...
+		be empty [=control message queue=]s.
 
 		2. Set the internal slot <dfn attribute
 			for="BaseAudioContext">[[current frame]]</dfn> of the
 			{{BaseAudioContext}} to 0.
+	</div>
 
-		3. Let <var>RenderResult</var> be false.
+	1. Let <var>RenderResult</var> be false.
```
Comment (rtoy) [inline index.bs]: Maybe wrap `<code>` around "false"?  Do the same for true too then.

File: index.bs [hunk 4]
```diff
@@ -10552,29 +10552,27 @@ of <a href="#rendering-a-graph">rendering a graph</a>.
...
-			<var>Q<sub>rendering</sub></var> and <var>Q</var>.
+<div id="rendering-initialization" algorithm="initialize rendering loop">
+	The following step MUST be performance once before the rendering loop starts.
 
-		2. Set the internal slot <dfn attribute
-			for="BaseAudioContext">[[current frame]]</dfn> of the
-			{{BaseAudioContext}} to 0.
+	1. Set the internal slot <dfn attribute
+		for="BaseAudioContext">[[current frame]]</dfn> of the
+		{{BaseAudioContext}} to 0. Also set {{BaseAudioContext/currentTime}} to 0.
```
Comment (rtoy) [inline index.bs]: Initialize currentTime as well, otherwise it's undefined until the first render happens.

File: index.bs [hunk 5]
```diff
@@ -10552,29 +10552,27 @@ of <a href="#rendering-a-graph">rendering a graph</a>.
...
 operation such as resolution of {{Promise}}s in the {{AudioWorkletGlobalScope}}.
 
-<div id="rendering-a-graph" algorithm="rendering a graph">
-	1. Prepare the rendering.
-
-		1. Let <var>Q<sub>rendering</sub></var> be an empty [=control message
-			queue=]. [=Atomically=] [=swap=]
-			<var>Q<sub>rendering</sub></var> and <var>Q</var>.
+<div id="rendering-initialization" algorithm="initialize rendering loop">
+	The following step MUST be performance once before the rendering loop starts.
```
Comment (chrisguttandin) [inline index.bs]: Sorry for barging in, but shouldn't it say "... MUST be performed once ..."?

File: index.bs (hunk 4 above)
Comment (hoch) [reply index.bs]: This is done. Or did you want something different?

File: index.bs (hunk 5 above)
Comment (hoch) [reply index.bs]: The autocompletion ruins it again... the fix is on the way. :)

**Pull request merged**